### PR TITLE
Tickets/dm 36722

### PR DESCRIPTION
**Pull request by iagaponenko**

### CODE REVIEW

**Review by jgates108 — APPROVED** [review]
Looks good, just minor comments.

File: src/replica/DatabaseServicesMySQL.cc [hunk 1]
```diff
@@ -1691,13 +1694,13 @@ TransactionContribInfo DatabaseServicesMySQL::createdTransactionContrib(
...
-                          loadTime, TransactionContribInfo::status2str(status), info.tmpFile, numWarnings,
-                          numRowsLoaded, failed ? info.httpError : 0, failed ? info.systemError : 0,
-                          failed ? info.error : string(), (failed ? info.retryAllowed : false) ? 1 : 0));
+        queries.emplace_back(_g.insert(
+                "transaction_contrib", Sql::NULL_, info.transactionId, info.worker, info.database, info.table,
+                info.chunk, info.isOverlap ? 1 : 0, info.url, info.async ? "ASYNC" : "SYNC", info.maxRetries,
+                numFailedRetries, numBytes, numRows, createTime, startTime, readTime, loadTime,
+                TransactionContribInfo::status2str(status), info.tmpFile, numWarnings, numRowsLoaded,
+                failed ? info.httpError : 0, failed ? info.systemError : 0, failed ? info.error : string(),
+                (failed ? info.retryAllowed : false) ? 1 : 0));
```
[the remark this answers is not in the document]
Comment (iagaponenko) [reply src/replica/DatabaseServicesMySQL.cc:1703]: Agree

File: src/replica/IngestRequest.cc [hunk 2]
```diff
@@ -177,6 +178,55 @@ shared_ptr<IngestRequest> IngestRequest::test(TransactionContribInfo const& cont
...
+    swap(retry.startTime, contrib.startTime);
+    swap(retry.readTime, contrib.readTime);
+    swap(retry.tmpFile, contrib.tmpFile);
+    swap(retry.httpError, contrib.httpError);
+    swap(retry.systemError, contrib.systemError);
+    swap(retry.error, contrib.error);
+
+    contrib.status = TransactionContribInfo::Status::IN_PROGRESS;
+    contrib.async = async;
+    contrib.retryAllowed = false;
```
[the remark this answers is not in the document]
Comment (iagaponenko) [reply src/replica/IngestRequest.cc:219]: Good suggestion! Will implement it.

File: src/replica/IngestRequest.h [hunk 3]
```diff
@@ -123,6 +123,33 @@ class IngestRequest : public std::enable_shared_from_this<IngestRequest>, public
...
+     * @param contribId A unique identifier of an existing contribution request.
+     * @param async The processing mode to be set at the request for bookkeeping purposes.
+     * @throw std::invalid_argument For non-existing request, or incorrect values of
+     *   the input parameters.
+     * @return A newly created instance of the request object.
+     */
+    static std::shared_ptr<IngestRequest> retry(std::shared_ptr<ServiceProvider> const& serviceProvider,
+                                                std::string const& workerName, unsigned int contribId,
+                                                bool async);
+
```
[the remark this answers is not in the document]
Comment (iagaponenko) [reply src/replica/IngestRequest.h:152]: Agreed

File: src/replica/IngestRequest.cc [hunk 4]
```diff
@@ -177,6 +178,55 @@ shared_ptr<IngestRequest> IngestRequest::test(TransactionContribInfo const& cont
...
+                "contribution id=" + to_string(contribId) + " is not in state " +
+                TransactionContribInfo::status2str(TransactionContribInfo::Status::READ_FAILED) +
+                ", the actual state is " + TransactionContribInfo::status2str(contrib.status) + ".");
+    }
+
+    // Move counters and error status codes from the contribution object
+    // into the retry. The corresponding fields of the contribution objects
+    // will get reset to the initial values (which are the same as in the default
+    // constructed retry object).
+    TransactionContribInfo::FailedRetry retry;
```
[the remark this answers is not in the document]
Comment (iagaponenko) [reply src/replica/IngestRequest.cc:207]: Sounds good to me

**Pull request merged**